### PR TITLE
Fix GPT-2 Flash Attention 2 generation with left-padding

### PR DESCRIPTION
**Pull request by Abdennacer-Badaoui**
# What does this PR do?

Fixes Flash Attention 2 generation with left-padding in GPT-2 by ensuring `is_causal=True` is set even when padding masks are provided.

Flash Attention 2 handles causal masking and padding independently, but the original code incorrectly disabled causal masking when any attention mask was present.

Fixes: `test_flash_attn_2_generate_padding_left`

### CODE REVIEW

**Comment by Abdennacer-Badaoui** [comment]
@vasqu , ready for review :)

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 1]
```diff
@@ -245,6 +245,11 @@ def forward(
 
         is_causal = attention_mask is None and query_states.shape[-2] > 1 and not is_cross_attention
 
+        # For flash attention backends, we must keep causal behavior even when a 2D padding mask is provided
+        # (flash attention uses `is_causal` for the triangular mask and expects an optional 2D key padding mask)
+        if self.config._attn_implementation in {"flash_attention_2", "flash_attention_3"} and not is_cross_attention:
+            is_causal = query_states.shape[-2] > 1
```
Comment (Cyrilvallez) [inline src/transformers/models/gpt2/modeling_gpt2.py]: I believe we can greatly simplify the check in the end, such as a simple
```suggestion
        is_causal = not is_cross_attention
```
if I'm not mistaken

**Review by Cyrilvallez — DISMISSED** [review]
Makes sense, fa should know that it's indeed causal! cc @vasqu here as well!

However, I think we can even completely remove the shape check, as sdpa checks it internally to make sure of the alignment, and fa should align the mask correctly already if I remember correctly.

File: src/transformers/models/gpt2/modeling_gpt2.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/gpt2/modeling_gpt2.py]: We should do this during init instead. It's probably very old code so the interface wasn't as established at that time. These edge cases should already be handled be the sdpa interface and we should not check for the same on general modeling code

Comment (vasqu) [reply src/transformers/models/gpt2/modeling_gpt2.py]: Relevant code part https://github.com/huggingface/transformers/blob/1c9077f66dc285dd32aa2410bed8acc680d64475/src/transformers/integrations/sdpa_attention.py#L69-L82

**Review by vasqu — COMMENTED** [review]
I'm with Cyril here, let's simplify the forward, essentially removing the whole is_causal logic during the forward and only change it within in the init based on if we are cross attn or not

**Comment by vasqu** [comment]
All the is_causal logic is correctly handled by the integrations; we just need to make sure we define it correctly on init

**Comment by Abdennacer-Badaoui** [comment]
Makes sense. Thanks @vasqu, @Cyrilvallez. Pushing the changes now.

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 2]
```diff
@@ -262,7 +260,7 @@ def forward(
                 value_states,
                 attention_mask,
                 dropout=self.attn_dropout.p if self.training else 0.0,
-                is_causal=is_causal,
+                is_causal=self.is_causal,
```
Comment (vasqu) [inline src/transformers/models/gpt2/modeling_gpt2.py]: We should be able to remove this line as well, the integrations grep it via the passed self (module)

**Comment by vasqu** [comment]
run-slow: decision_transformer, gpt2

**Comment by vasqu** [comment]
run-slow: decision_transformer, gpt2

**Review by vasqu — APPROVED** [review]
Thx a lot, looking good now!

**Comment by vasqu** [comment]
@Abdennacer-Badaoui can you run the fa tests locally and confirm that those work? Seems like our CI doesn't have FA installed anymore so just wanna double check

**Comment by Abdennacer-Badaoui** [comment]
@vasqu, I tested the FA failing tests for which I opened the PR. they’re all passing.

**Comment by vasqu** [comment]
Thx a lot :heart:
